---
output:
  html_document: default
  pdf_document: default
---
## Lab 2: Identifying population structure from genomic data


Name:
```
Write your name here
```

**Due Date: **


### Goals:
This exercise is meant to give you tools for visualizing population structure in genomic datasets. Additionally, you will test whether some individuals have recent admixture from different ancestries and whether individuals within a population are closely related to each other. Some of the work is in R. You will learn how to make your own ADMIXTURE and PCA graphics with the code provided

## Task 1. Plot and Interpret PCA:
As always please ssh into the tadpole HPCC; navigate to your personal directory in the student folder, then make a directory called `lab2`.

Please use the `cp` command to copy the `Ethiopians_chr1_Update_Sex` binary files you created in the `Lab1` folder from the last assignment to the`Lab_2` directory you just created. 

Now run the following command to produce a QCed set to use from this point onward:

```
plink --bfile Ethiopians_chr1_Update_Sex --geno 0.02 --maf 0.05 --hwe 0.001 --make-bed --out Ethiopians_Update_Sex_chr1_geno02_maf05_hwe001

```

You can run a PCA in plink by using the --pca command (see plink website and Ch. 9 of your book for more details). In this instance, we are using the plink command below to run the first 4 principal components on the Ethiopians_Chr1 dataset. Run this command on the binary file dataset.

```
plink --bfile Ethiopians_Update_Sex_chr1_geno02_maf05_hwe001 --pca 4 --out EthiopianChr1_pcs4
```

## Question 1 [ pts]

Look at the .eigenvec file. Describe the file contents. What is contained in the row and the columns? You may need to reference this link (https://www.cog-genomics.org/plink/2.0/formats#eigenvec)

```
(Type your answer here)
```

Now we will plot the PCA. Open a new terminal window (you can do this with Command+T (CTRL+T for windows) the same way you would open a new window on the internet). DON’T login to tadpole. Instead use pwd to see where you are on your local computer; this is your current working directory. In order to move a file from a remote host (tadpole) to your local host (your computer) we can use the ```scp``` command. NOTE this is all in one command line. Fill in the login and file names with yours. Recall that “.” (punto) means your current working directory on your computer, just like you practiced in Unix Basics (you can change this if you want to download things elsewhere).

```
scp loginName@tadpole.genomecenter.ucdavis.edu:/path/to/eigenvec/file .
```

Open R. In R console, set your working directory to where you downloaded the file.

```
setwd("path_to_my_files")
```
Then we will install a package called ggplot to visualize the data
> NOTE: You only need to do this if you don't already have the ggplot2 package. If you already have it, you can skip to the library() command

if you don't already have ggplot installed in your Rstudio from previous classes or work copy paste the next line directly into your R console 
```
install.packages("ggplot2")

```

Now we will read in the data to a table we call “data” and relabel the columns to be easier to read. Run the following code below in your R console

```{r, eval=TRUE, echo=TRUE}
library(ggplot2)
columns<-c("population", "sampleID", "PC1", "PC2", "PC3", "PC4")
data<- read.table("EthiopianChr1_PCs4.eigenvec", sep = "",  header=F, col.names=columns)
```

## Question 2 [4 pts]

Add your plots of PC1 vs PC2, and PC3 vs PC4 below. This can be done by pasting your R plotting code in the section below that specifies eval=TRUE. The eval=TRUE specification will tell R to run the code (to generate the plot) when you knit the file, and the echo=TRUE will specify that your code will be shown in addition to being evaluated/run in the knitted file. However, You can run the code to visualize your plot before you knit the final file by clicking the green triangle on the top right of the code block.
> Note: Your evaluated code below will need to include your setwd("[path_to_files]/") and library(ggplot2) in order to evaluate 

Generate your PCA plots in R below

PC1 vs PC2
```{r, eval=TRUE, echo=TRUE}

ggplot(data, aes(x=PC1,y=PC2, color=population))+geom_point()
```
PC3 vs PC4
```{r, eval=TRUE, echo=TRUE}
ggplot(data, aes(x=PC3,y=PC4, color=population))+geom_point()


```

Do we observe population structure in this dataset? Explain. Which populations are most closely related to each other? Which are most distant?

```
(Type your answer here)
```

## Question 3 [2pts]

Do you learn any additional information by looking at PC3 vs PC4 (as opposed to just the first two PCs?

```
(Type your answer here)
```

## Question 4 [2pts]

Which command do you use to remove old files? How about removing multiple files at once?

```
(Type your answer here)
```

> Note: Do not miss this step below! It is important that you copy over this sorted dataset. If the fam file is not sorted, then pong (future step) will produce an admixture plot that is not grouped by population, and we don't want that!

Copy the ```/share/ant157/data/week4/Ethiopians_All_UpdateSex_Sort.* ``` plink dataset into your new directory.

## Question 5 [2pts]

How many chromosomes and SNPs are in this file?

> Hint: To find the number of chromosomes one way is to use the `awk` command on the first column of the relevant plink file and pipe (|) it to the `uniq` command

```
(Type your answer here)
```

Now let’s create a new binary dataset with fewer correlated SNPs using the sorted dataset bfiles you just copied over. Let’s remove SNPs that have an r<sup>2</sup> (r-squared) of 0.5 or higher within a 50 kb window. --indep-pairwise takes 3 parameters, window-size, the number of variants by which the window slides down the chr and the cutoff for the degree of correlation (r<sup>2</sup>).  

```
plink --bfile Ethiopians_All_UpdateSex_Sort --indep-pairwise 50 40 0.5 --out EthiopiansAll_LD050
```

Running the code above creates EthiopiansAll_LD050.prune.in and EthiopiansAll_LD050.prune.out file. We will now use the `--extract` flag to make a new set of binary files using the code below:
```
plink --bfile yourdataset --extract EthiopiansAll_LD050.prune.in --make-bed --out EthiopiansAll_LDpruned050
```

## Question 6 [4 pts]

How many SNPs are left in your LD-thinned (pruned) dataset?

```
(Type your answer here)
```

Re-run PCA on this dataset. Add the PCA plot (PC1 vs PC2) into the space below via the same method as above. Is this any different from the PCA on just chr1? Why?

Put your plotting code below to generate your image
```{r, eval=TRUE, echo=TRUE}

columns<-c("population", "sampleID", "PC1", "PC2", "PC3", "PC4")
data<- read.table("EthiopiansAll_LDpruned050_pcs4.eigenvec", sep = "",  header=F, col.names=columns)
ggplot(data, aes(x=PC1,y=PC2, color=population))+geom_point()


```
```
(Type your answer here)
```

## Task 2. Start a preliminary run of ADMIXTURE.

You will be submitting a shell script to run `ADMIXTURE` in the background while you work on other things. Shell scripts end in .sh suffix. It can take time for `ADMIXTURE` to run, especially for higher Ks. We’re going submit a script to run `ADMIXTURE` for K = 2,3,4,5 and 6 different ancestral clusters (i.e., at the end you’ll have run `ADMIXTURE` 5 times).


The first step is to run admixture on your dataset of interest (i.e. your sorted, LD-pruned Ethiopian dataset) using the example script below. The class will submit the script as a slurm job using:

```
sbatch yourscriptname.sh
```

Where you replace the script name with whatever yours is called.

The script looks like this:

```
#!/bin/bash
#SBATCH --job-name="myjob" # Job name -you can change "myjob" to something different
#SBATCH --nodes=1 # single node, anything more than 1 will not run
#SBATCH --ntasks=1 # equivalent to cpus -don't change
#SBATCH --time=1-00:00:00 # expected time of completion in days-hours:minutes:seconds -dont'change
#SBATCH -p production # partition to submit to, dont'change
#SBATCH --account=ant157 #-don't change
#SBATCH --mail-type=ALL #don't change

# Replace your email here!
#SBATCH --mail-user=you@gmail.com


module load admixture
for K in 2 3 4 5 6
do
  admixture -s 35 <path to EthiopiansAll_LDpruned050.bed> $K | tee log${K}.out
done
```


The output of admixture should be a .P and .Q file for every iteration of "K" that you selected, as well as the log files for every K iteration.

Make sure you change the script to give your job a unique name and call the correct files. The `-s` flag simply specifies a random number to seed the program. You can change this random number if you want (right now it is 35). Remember to write in the correct path to your `.bed` plink file.

edit the code chunk above, making all the required changes, use `nano` to create a file called `admixture_script.sh` then copy paste the edited code chunk into this file.

Submit this job to the cluster by running `sbatch yourscriptname.sh` . We’ll get back to plotting the output later. Check back in a few minutes to make sure that your job is running correctly and that the output files were produced for K=2. The whole thing could take 6-8 hours to finish. When it’s done, you should see your output files – a set of .P, .Q and log files.

You can check the status of your job with
```
squeue -u <username>
```

## Task 3. Plot and Interpret ADMIXTURE:

### 1. Running Pong to plot admixture runs

Pong requires an input file with three **tab-delimited** columns:

- Column 1: A unique run ID for each Q matrix (must begin with a letter, can include numbers and underscores)   
- Column 2: The K value for the Q matrix  
- Column 3: The file path to the Q matrix


To begin, create this “filemap.txt” file in the same directory as your P and Q matrices.

Example filemap.txt:

```
k2  2 /share/ant157/students/YourName/EthiopiansAll_LDpruned050.2.Q
k3  3 /share/ant157/students/YourName/EthiopiansAll_LDpruned050.3.Q
k4  4 /share/ant157/students/YourName/EthiopiansAll_LDpruned050.4.Q
k5  5 /share/ant157/students/YourName/EthiopiansAll_LDpruned050.5.Q
k6  6 /share/ant157/students/YourName/EthiopiansAll_LDpruned050.6.Q
```

In order to have population labels on the plot, Pong also requires a file with a single column containing the population IDs for each individual in order. They refer to it as the “ind2_pop” file in the [manual](http://brown.edu/Research/Ramachandran_Lab/files/pong/pong-manual.pdf)

This information can be found in the `Ethiopians_All_UpdateSex_Sort.fam` file, in the first column. You can use this bash command to extract the first column and save it in a new file.
```
cut -d " " -f1 Ethiopians_All_UpdateSex_Sort.fam > ind2_pop.txt
```

Pong works by opening up a connection to a localhost on a web browser in order to produce an interactive visualization. In order to run pong on the remote server (tadpole), we need to set up port forwarding from your laptop’s port 4000 to the remote cluster’s port 4000.

> Note: only one person at a time can be connecting via a given port number like 4000. So, you will be assigned your OWN number to connect to as seen below:

Carlos: 	4001
Juan: 	4002
Juliana: 	4003
Estefania: 	4004
César: 	4005
Fabian: 	4006
Yuqi:	4007

Follow these steps below in order to set up port forwarding and run pong:

1. Open up another terminal window, but do not connect to tadpole. Run the command below, replacing `4000` with the number you are assigned. On a mac, you can use command+T to open another terminal tab.
    ```
    ssh -N -L 4000:localhost:4000 username@tadpole.genomecenter.ucdavis.edu
    ```
    You will be prompted to put in your password. The command will then hang: This is normal. **Do not close the terminal window**, just let the command hang there.

2. Return to your tadpole terminal window

3. Navigate to the directory containing your P and Q matrices, and the filemap.txt and mapfile you created. To run pong, simply use:
```
module load pong
pong -m filemap.txt -i ind2_pop.txt --port YourAssignedNumber
```


4. You should see an output that looks like this (but it might take a second...do not panic!)

```
-------------------------------------------------------------------
                            p o n g
      by A. Behr, K. Liu, T. Devlin, G. Liu-Fang, and S. Ramachandran
                       Version 1.4.7 (2016)
-------------------------------------------------------------------
-------------------------------------------------------------------

Parsing input and generating cluster network graph
Matching clusters within each K and finding representative runs
For K=5, there is 1 mode across 1 run.
Matching clusters across K
Finding best alignment for all runs within and across K
match time: 0.00s
align time: 0.00s
total time: 0.03s
```

5. Open up your web browser of choice (Chrome, Safari etc) and type in http://localhost:4000 (subbing your assigned number for 4000)

6. Pong should open up an interactive visualization in your web browser.

7. When you are done, simply close the terminal tab where the port forwarding ssh command is hanging and the connection will close.


## Question 7 [4 pts]

 If you were to create a population structure plot using ADMIXTURE for k=3, which ancestral clusters would you expect to observe? How about at k=4? Write down your expectations.
  > Hint: you may need to google the population names and do a little research on the groups. Feel free to discuss this with your classmates 

K = 3:
 ```
(Type your answer here)
 ```
K = 4
 ```
(Type your answer here)
 ```

## Question 8 [3 pts]

 Save the image of your `ADMIXTURE` plots for K=2 through K=6 and upload them with your homework document/answers. Interpret each plot, describing how the changes you observe as you increase K. If you don’t already know, you should research the geographic locations of all the populations you plotted.

```
(Type your answer here. Don't forget to turn in your plots too! Please list the name of the file for these plots.)
```


## Question 9 [2 pts]

 Take a look at K=6. You should see that some individuals no longer look ‘admixed’, but instead have their own unique color that they share with a few other individuals in their population. Why do you think you see these unique clusters? Explain

 ```
(Type your answer here)
 ```

## Question 10 [2pts]

 Look at the K=2 output files from `ADMIXTURE`. At the bottom of the log file, there will be an estimate of Fst between the two ancestral populations. What is this number? Is differentiation between the two ancestral populations very high, very low or moderate, in the context of humans?

 ```
(Type your answer here)
 ```

## Task 5. Compare PCA and ADMIXTURE

  let's ask how PCA and ADMIXTURE are complementary. Navigate to /share/ant157/data/1kg and copy the  1kg_EU_QC binary files to your `lab2` folder. Then look at the 1kg_EU_QC.fam file.
  
## Question 11 [1pt]
  Which populations are listed in this fam file?

 ```
(Type your answer here)
 ```
 
Navigate back to your own directory and perform the LD thinning procedure as you did above, on this new dataset. Write the new dataset (HINT: "--make-bed --out") commands to your directory. Now run PCA for PCs 1-4 and ADMIXTURE K=4 on this LD-thinned dataset. Copy your plots below.

PC1 vs PC2
```{r, eval=TRUE, echo=TRUE}


```
PC3 vs PC4
```{r, eval=TRUE, echo=TRUE}


```

K = 4
 ```
(Type your answer here)
 ```

## Question 12 [4 pts]
  Compare and contrast the PCA and the ADMIXTURE plots. How are they similar? different? Explain.

 ```
(Type your answer here)
 ```