Course: D43, Human Population & Computational Genetics

Course Instructor: Oshiomah Oyageshio (pronounced Oh-shoh-mah)

Course Zoom Link:

Course Time: Wednesdays 5:00-6:30 pm (lecture), Thursdays 5:00-7pm (computer lab)

Course Duration: April 4 – May 30

Office Hours: TBD

E-mail: [oyageshio@ucdavis.edu](mailto:oyageshio@ucdavis.edu) **\*\* Please include “D43” in the subject Line of the e-mail**

COURSE DESCRIPTION:

This is an online course that emphasizes population genetic theory as integrated into a practical hands-on computational experience. Students will build on a theory-based lecture to gain a detailed understanding of population genetics by extensive data analysis. The focus of the lab component is on human genetic/genomic example datasets.

Course Objectives:

At the end of the course, students will know how to:

1. manipulate human genetic SNP datasets using command-line (UNIX) procedures
2. plot results from data analysis in R
3. perform an admixture analysis on genome-wide SNP data
4. perform a genome-wide association study (GWAS) for quantitative and disease traits
5. calculate polygenic risk scores from GWAS results
6. use common genome browsers to identify pertinent information about human SNPs and their locations relative to known genes
7. use genetic data to identify cryptic relatives in human genetic datasets and confirm reported pedigrees
8. perform an analysis of heritability for a complex trait

Population Genetic Concepts:

Hardy-Weinberg equilibrium, population structure, F<sub>st</sub>, admixture, ancestry, populations, effective population size, haplotypes, linkage disequilbrium, identity by descent vs state, mutation rate, genetic drift, migration rates, heritability, inbreeding, neutral theory, positive and negative selection, complex traits, GWAS, ascertainment bias, polygenic scores

TEACHING SCHEDULE: Subject to change-- stay tuned

**APRIL 2024**

| Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
| --- | --- | --- | --- | --- | --- | --- |
| 31  | 1   | 2   | 3   | 4   | 5   | 6   |
|     |     |     | **Lecture;No class on this day** | Lab: _UNIX and scripting tutorial_ |     |     |
| 7   | 8   | 9   | 10  | 11  | 12  | 13  |
|     |     |     | Human genomics introduction – variants, heterozygosity, populations | Lab |     |     |
| 14  | 15  | 16  | 17  | 18  | 19  | 20  |
|     |     |     | Allele frequencies -- array quality control, hardy-weinberg equilibrium | Lab |     |     |
| 21  | 22  | 23  | 24  | 25  | 26  | 27  |
|     |     |     | Population structure – admixture, PCA | Lab |     |     |
| 28  | 29  | 30  |     |     |     |     |
|     |     |     |     |     |     |     |

**MAY 2024**

| Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
| --- | --- | --- | --- | --- | --- | --- |
|     |     |     | 1   | 2   | 3   | 4   |
|     |     |     | Genome-wide association studies – neutral traits, disease traits | Lab |     |     |
| 5   | 6   | 7   | 8   | 9   | 10  | 11  |
|     |     |     | Polygenic risk scores, linkage disequilbrium | Lab |     |     |
| 12  | 13  | 14  | 15  | 16  | 17  | 18  |
|     |     |     | Inbreeding -- F coefficients, runs of homozygosity | Lab |     |     |
| 19  | 20  | 21  | 22  | 23  | 24  | 25  |
|     |     |     | Pedigrees – identity by descent, haplotypes | Lab |     |     |
| 26  | 27  | 28  | 29  | 30  | 31  |     |
|     |     |     | Final Lecture: TBD | Final Lab: |     |     |

Course Material and Required Reading

- Required - An Introduction to Statistical Genetic Data Analysis, by Melinda C. Mills, Nicola Barban, Felix C. Tropf, 2020 MIT Press
- Readings from the primary literature may be assigned. If included, journal articles will be made freely available via pdf to all students.

READING SCHEDULE:

Specific textbook pages for reference will be given for relevant problem sets. Primary literature TBD.

Assignments and Tests:

- Five homework assignments based on class lecture and the textbook will be assigned.
- **OPTIONAL/TBD:** A group project or independent project will be due on the last week. Each group/person will be offered access to a genomic dataset. In consultation with the instructor, they will design a research question pertaining to quantitative trait architecture or demographic inference. Students will work collaboratively to implement different analyses and interpret their results. A short written project report will be handed in with annotated student contributions.
